db/blob/blob_source.cc

```diff
@@ -0,0 +1,295 @@
+
+  // First, try to get the blob from the cache
+  //
+  // If either blob cache is enabled, we'll try to read from it.
+  Status s;
+  Slice key;
+  if (blob_cache_ != nullptr) {
+    key = cache_key.AsSlice();
+    s = GetBlobFromCache(key, blob_cache_.get(), blob_entry, wait);
+    if (blob_entry->GetValue()) {
+      if (prefetch_buffer) {
+        // Update the blob details so that PrefetchBuffer can use the read
+        // pattern to determine if reads are sequential or not for prefetching.
+        // It should also take in account blob read from cache.
+        prefetch_buffer->UpdateReadPattern(
+            offset, blob_entry->GetValue()->size(),
+            read_options.adaptive_readahead /* decrease_readahead_size */);
+      }
+    }
+  }
+
+  assert(s.ok() || blob_entry->GetValue() == nullptr);
+
+  return s;
+}
+
+Cache::Handle* BlobSource::GetEntryFromCache(
+    const CacheTier& cache_tier, Cache* blob_cache, const Slice& key,
+    const bool wait, const Cache::CacheItemHelper* cache_helper,
+    const Cache::CreateCallback& create_cb, Cache::Priority priority) const {
+  assert(blob_cache);
+
+  Cache::Handle* cache_handle = nullptr;
+
+  if (cache_tier == CacheTier::kNonVolatileBlockTier) {
+    cache_handle = blob_cache->Lookup(key, cache_helper, create_cb, priority,
+                                      wait, statistics_);
+  } else {
+    cache_handle = blob_cache->Lookup(key, statistics_);
+  }
+
+  return cache_handle;
+}
+
+Status BlobSource::InsertEntryToCache(
+    const CacheTier& cache_tier, Cache* blob_cache, const Slice& key,
+    const Cache::CacheItemHelper* cache_helper, const Slice* blob,
+    size_t charge, Cache::Handle** cache_handle,
+    Cache::Priority priority) const {
+  Status s;
+
+  if (cache_tier == CacheTier::kNonVolatileBlockTier) {
+    s = blob_cache->Insert(
+        key, const_cast<void*>(static_cast<const void*>(blob->data())),
+        cache_helper, charge, cache_handle, priority);
+  } else {
+    s = blob_cache->Insert(
+        key, const_cast<void*>(static_cast<const void*>(blob->data())), charge,
+        cache_helper->del_cb, cache_handle, priority);
+  }
+
+  return s;
+}
+
+Status BlobSource::GetBlobFromCache(const Slice& cache_key, Cache* blob_cache,
+                                    CachableEntry<Slice>* blob,
+                                    bool wait) const {
+  assert(blob);
+  assert(blob->IsEmpty());
+  const Cache::Priority priority = Cache::Priority::LOW;
+
+  // Lookup uncompressed cache first
+  if (blob_cache != nullptr) {
+    assert(!cache_key.empty());
+    Cache::Handle* cache_handle = nullptr;
+    cache_handle = GetEntryFromCache(
+        lowest_used_cache_tier_, blob_cache, cache_key, wait,
+        nullptr /* cache_helper */, nullptr /* create_db */, priority);
+    if (cache_handle != nullptr) {
+      Slice value(static_cast<char*>(blob_cache->Value(cache_handle)));
+      blob->SetCachedValue(&value, blob_cache, cache_handle);
```

Comment:
Storing a pointer to this stack-based `value` in the `CachableEntry` is not safe, since `value` is about to get destroyed.

Reply:
Fixed.